tornado-cash-monitor/src/agent.js

```diff
@@ -0,0 +1,174 @@
+    } = config;
+
+    data.developerAbbreviation = developerAbbreviation;
+    data.protocolName = protocolName;
+    data.protocolAbbreviation = protocolAbbreviation;
+    data.observationIntervalInBlocks = observationIntervalInBlocks;
+    data.iface = new ethers.utils.Interface(abi);
+
+    // get the address names specified in the config
+    const addressNames = Object.keys(addressList);
+    if (addressNames.length === 0) {
+      throw new Error('Must supply at least one address to watch');
+    }
+
+    data.addressesToMonitor = [];
+    addressNames.forEach((addressName) => {
+      const info = {
+        name: addressName,
+        address: addressList[addressName].address,
+        type: addressList[addressName].type,
+        severity: addressList[addressName].severity,
+      };
+
+      data.addressesToMonitor.push(info);
+    });
+
+    // create an object to hold addresses that have been identified as having interacted with a
+    // Tornado Cash Proxy
+    data.suspiciousAddresses = {};
+    /* eslint-enable no-param-reassign */
+  };
+}
+
+function provideHandleTransaction(data) {
+  return async function handleTransaction(txEvent) {
+    const findings = [];
+    const {
+      developerAbbreviation,
+      protocolName,
+      protocolAbbreviation,
+      observationIntervalInBlocks,
+      iface,
+      addressesToMonitor,
+    } = data;
+
+    // check to see if the given transaction includes deposit/withdraw calls from a tornado cash
+    // proxy
+    let addressesOfInterest = TORNADO_CASH_ADDRESSES.map((address) => {
+      const filterResult = txEvent.filterFunction(
+        iface.format(ethers.utils.FormatTypes.full, address),
+      );
+
+      if (filterResult.length > 0) {
+        return txEvent.from;
+      }
+
+      return '';
+    });
+
+    // filter out any empty strings
+    addressesOfInterest = addressesOfInterest.filter((address) => address !== '');
+
+    // for each address found to have interacted with a tornado cash proxy, add it to our
+    // suspiciousAddresses object and instantiate a number of blocks to watch the address for; if
+    // an address is already present in suspiciousAddresses then simply restart its block timer
+    addressesOfInterest.forEach((address) => {
+      // eslint-disable-next-line no-param-reassign
+      data.suspiciousAddresses[address] = {
+        blockAdded: txEvent.blockNumber,
+      };
+    });
+
+    // iterate over the list of suspiciousAddresses and check to see if any address can be removed
+    const addressesToRemove = [];
+    Object.keys(data.suspiciousAddresses).forEach((address) => {
+      const currBlock = txEvent.blockNumber;
+      const { blockAdded } = data.suspiciousAddresses[address];
+      if ((currBlock - blockAdded) >= observationIntervalInBlocks) {
+        // block is older than observationIntervalInBlocks and can be removed from
+        // suspicousAddresses
+        addressesToRemove.push(address);
+      }
+    });
```

Comment:
shouldn't this be removed after the findings, otherwise the observationIntervalInBlocks is off by one

Reply:
I'll change the `>=` to just `>` so that it isn't off by one. Additionally, I moved the check to before generating findings since in the (unlikely) event that the agent misses blocks we don't want to generate findings for suspicious addresses that are over the observation interval.